### ISSUE TITLE
Blocks does not work with nested array structure

### ISSUE THREAD

**Issue by yahelnachum** [jsonata-js/jsonata#332]
# Summary
In the [jsonata playground](http://try.jsonata.org/) I was trying out [parenthesized expressions and blocks](http://docs.jsonata.org/composition#parenthesized-expressions-and-blocks) and came across some inconsistent output. I used blocks to turn a selection into an array so that I could pick a single index. Then when I tried adding a key to step into that single index it seemed to revert back to a selection.

# More Details

JSON Input 1:
```
[
  {
    "phone": [
      {"number" : 0}
    ]
  },
  {
    "phone": [
      {"number" : 1}
    ]
  },
  {
    "phone": [
      {"number" : 2}    
    ]
  }
]
```

Then I tried a series of JSONata expressions

----
Expression 1:
```
phone
```

Actual Output 1:
```
[
  {"number":0},
  {"number":1},
  {"number":2}
]
```

---
Expression 2:
```
(phone)[0]
```

Actual Output 2:
```
{
  "number": 0
}
```
----
Expression 3:
```
(phone)[0].number
```

Actual Output 3:
```
[
  0,
  1,
  2
]
```

----
I would have expected the output of `Expression 3` to be the following given that I was able to select a single index in `Expression 2`.

Expected Output 3:
```
0
```

This seems to me like a bug in the code somewhere, but if there is an explanation for this behavior then please let me know what that is. Thanks in advance for reading!

**Comment by yahelnachum** [comment]
My co-worker was trying some other inputs and got the expected behavior
JSON input 2:
```
{
  "account": [
    {
      "phone": [
        {"number" : 0}
      ]
    },
    {
      "phone": [
        {"number" : 1}
      ]
    },
    {
      "phone": [
        {"number" : 2}    
      ]
    }
  ]
}
```

----
Expression 4:
```
(account.phone)[0].number
```

Actual Output 4:
```
0
```

----

So for some reason wrapping my `JSON input 1` in a JSON object gives me the correct behavior.

**Comment by yahelnachum** [comment]
Just tried another input with a nested array with the same `Expression 3` and it worked this time. Still not sure why its behaving differently.

JSON input 3:
```
[
  [
    {
      "phone": [
        {"number" : 0}
      ]
    },
    {
      "phone": [
        {"number" : 1}
      ]
    },
    {
      "phone": [
        {"number" : 2}    
      ]
    }
  ]
]
```

----
Expression 5:
```
(phone)[0].number
```

Actual Output 5:
```
0
```

**Comment by andrew-coleman** [comment]
Yes, the result of your expression 3 looks like a bug.

**Comment by andrew-coleman** [comment]
@yahelnachum - because the input data is an array, to access the first item, you can use `$[0]`.  So your expression 3 could be rewritten as
```
$[0].phone.number
```
which will return
```
0
```

**Comment by andrew-coleman** [comment]
This issue is related to how JSONata handles input arrays.  There is a lengthy discussion in #170 on this topic.  I'm currently prototyping a solution that uses explicit syntax to determine whether to map over the input array verses just use it as a single context value.

**Comment by andrew-coleman** [comment]
resolved by #363
